elastic/kibana
# [Reporting] Event Log integration

## Summary

Part of: https://github.com/elastic/kibana/issues/29281
Part of: https://github.com/elastic/kibana/issues/121468
Needed for: https://github.com/elastic/kibana/issues/34951

This PR provides a way to track aggregatable metrics of Reporting jobs, using the Kibana Event Log and ECS log data. As an initial change, this PR tracks metrics of Downloading CSV from a Dashboard saved search.

We want to help users understand why the failures happen, without them having to grep through the logs. We want users to find ways to save money, if they have more server power than they need.

Using detailed metrics we could make things happening inside of Reporting very clear to an administrator. They could spot if a report job failed because some particular user tried to make a PDF of a dashboard with 101 panels. Or they could identify if CSV exports are failing because Elasticsearch needs more time to keep the scroll context open.

This PR starts indexing a set of Reporting metrics. These were chosen as things that relate to a factor of monitoring or performance measuring:
 - Categorical information: the type of job and the type of app using the job
 - Affects performance: the number of panels in a dashboard or the number of CSV rows exported
 - Is a result of performance: the amount of time the job is waiting in the queue, the time it takes to complete the job


Field | Value Type
-- | -- 
kibana.uuid | UUID of the Kibana instance (automatically provided by Event Log service)
kibana.reporting.jobType | `csv_searchsource_immediate`, `pngV2`, etc
kibana.reporting.contentType | `csv`, `pdf`, `png`
~~kibana.reporting.attempt~~ | ~~1-3~~
~~kibana.reporting.status~~ | ~~`completed`, `error`~~
~~kibana.reporting.queuedTime~~ | ~~Timestamp the report was queued (`job.created_at`)~~
~~kibana.reporting.completedTime~~ | ~~Timestamp the report was completed (`job.completed_at`)~~
~~kibana.reporting.settings.csv.maxSizeBytes~~ | ~~Kibana’s `xpack.reporting.csv.maxSizeBytes` setting~~
~~kibana.reporting.settings.csv.scrollSize~~ | ~~Kibana’s `xpack.reporting.csv.scroll.size` setting~~
~~kibana.reporting.settings.contentStream.chunkSize~~ | ~~Value determined from Elasticsearch’s `http.max_content_length` setting~~
kibana.reporting.byteSize | Number of total bytes of the report output
~~kibana.reporting.csv.numRows~~ | ~~Number of rows exported in the CSV~~
~~kibana.reporting.csv.scrollTime~~ | ~~median number of the duration of time between each call to scroll~~
~~kibana.reporting.csv.numColumns~~ | ~~number of fields selected in the saved search, or count of `*`~~
user.name | Username of logged-in user or null if Security is not enabled
event.kind |  `event`: for instantaneous action such as scheduling a report, saving a report.<br/>`metrics`: for action that has duration fields for timing, and execution metrics such as byteSize of the report.<br/>`error`: for handling execution failure
event.provider | `reporting`
event.action | `schedule-task` (kind: `event`)<br/>`claim-task` (kind: `event`)<br/>`execute-start` (kind: `event`)<br/>`execute-complete` (kind: `metrics`)<br/>`execute-error` (kind: `error`)<br/>`save-report` (kind: `event`)<br/>`retry` (kind: `event`)<br/>`fail-report` (kind: `event`)
event.id | Reporting Job ID, for correlating “queue” action with “execute” actions
event.timezone | Timezone of the user’s client (from report job params)
event.created | (For `execute-complete` actions) Timestamp that execution started
event.end | (For `execute-complete` actions) Timestamp that execution finished
event.duration | (For `execute-complete` actions) Time between execution start and complete
log.logger | `reporting`
log.level | `info`, `error`
error.message | (For `execute-error` actions) Message of the Error that was caught
task.id | ID of the `report:execute` task fired by Task Manager.

### References
 - ECS Guides:
    - https://www.elastic.co/guide/en/ecs/1.8/ecs-log.html
    - https://www.elastic.co/guide/en/ecs/1.8/ecs-event.html
    - https://www.elastic.co/guide/en/ecs/1.8/ecs-error.html
 - Event Log README: https://github.com/elastic/kibana/tree/7.16/x-pack/plugins/event_log

### Testing
1. Create an Data View based on `.kibana-event-log*`
2. Create a view in Discover of the Event Log data
   - Viewing data should require superuser permission, or a role that grants read access to the event log index
3. Run a few Download CSV jobs or report jobs
4. Ensure that reporting events are added to the log

### Checklist

Delete any items that are not applicable to this PR.
